example_10/hardware/rrbot.cpp

```diff
@@ -0,0 +1,280 @@
+      2);
+    return hardware_interface::CallbackReturn::ERROR;
+  }
+  // with exactly 1 command interface
+  for (int i = 0; i < 2; i++)
+  {
+    if (info_.gpios[i].command_interfaces.size() != 1)
+    {
+      RCLCPP_FATAL(
+        rclcpp::get_logger("RRBotSystemWithGPIOHardware"),
+        "GPIO component %s has '%ld' command interfaces, '%d' expected.",
+        info_.gpios[i].name.c_str(), info_.gpios[i].command_interfaces.size(), 1);
+      return hardware_interface::CallbackReturn::ERROR;
+    }
+  }
+  // and 3/1 state interfaces, respectively
+  if (info_.gpios[0].state_interfaces.size() != 3)
+  {
+    RCLCPP_FATAL(
+      rclcpp::get_logger("RRBotSystemWithGPIOHardware"),
+      "GPIO component %s has '%ld' state interfaces, '%d' expected.", info_.gpios[0].name.c_str(),
+      info_.gpios[0].state_interfaces.size(), 3);
+    return hardware_interface::CallbackReturn::ERROR;
+  }
+  if (info_.gpios[1].state_interfaces.size() != 1)
+  {
+    RCLCPP_FATAL(
+      rclcpp::get_logger("RRBotSystemWithGPIOHardware"),
+      "GPIO component %s has '%ld' state interfaces, '%d' expected.", info_.gpios[0].name.c_str(),
+      info_.gpios[0].state_interfaces.size(), 1);
+    return hardware_interface::CallbackReturn::ERROR;
+  }
+
+  return hardware_interface::CallbackReturn::SUCCESS;
+}
+
+hardware_interface::CallbackReturn RRBotSystemWithGPIOHardware::on_configure(
+  const rclcpp_lifecycle::State & /*previous_state*/)
+{
+  // BEGIN: This part here is for exemplary purposes - Please do not copy to your production code
+  RCLCPP_INFO(rclcpp::get_logger("RRBotSystemWithGPIOHardware"), "Configuring ...please wait...");
+  // END: This part here is for exemplary purposes - Please do not copy to your production code
+
+  // reset values always when configuring hardware
+  std::fill(hw_states_.begin(), hw_states_.end(), 0);
+  std::fill(hw_commands_.begin(), hw_commands_.end(), 0);
+  std::fill(hw_gpio_in_.begin(), hw_gpio_in_.end(), 0);
+  std::fill(hw_gpio_out_.begin(), hw_gpio_out_.end(), 0);
+
+  RCLCPP_INFO(rclcpp::get_logger("RRBotSystemWithGPIOHardware"), "Successfully configured!");
+
+  return hardware_interface::CallbackReturn::SUCCESS;
+}
+
+std::vector<hardware_interface::StateInterface>
+RRBotSystemWithGPIOHardware::export_state_interfaces()
+{
+  std::vector<hardware_interface::StateInterface> state_interfaces;
+  for (uint i = 0; i < info_.joints.size(); i++)
+  {
+    state_interfaces.emplace_back(hardware_interface::StateInterface(
+      info_.joints[i].name, hardware_interface::HW_IF_POSITION, &hw_states_[i]));
+  }
+
+  RCLCPP_INFO(rclcpp::get_logger("RRBotSystemWithGPIOHardware"), "State interfaces:");
+  hw_gpio_in_.resize(4);
+  size_t ct = 0;
+  for (size_t i = 0; i < info_.gpios.size(); i++)
+  {
+    for (auto state_if : info_.gpios.at(i).state_interfaces)
+    {
+      state_interfaces.emplace_back(hardware_interface::StateInterface(
+        info_.gpios.at(i).name, state_if.name, &hw_gpio_in_[ct++]));
```

Comment:
I see that for GPIOs we don't really have a standard interface name so were using whatever name is specified in the URDF. Would it make sense to create a new HW_IF_BOOLEAN interface type?
I found myself needing this interface also when controlling an actuator so I think it could be generally useful. Opinions?

Reply:
we discussed something related to this in the last WG meeting: GPIO is now only for semantic reasons, and is of type double. Supporting other datatypes (boolean, or even ROS msgs) is a repeating topic but no one made a sustainable suggestion on how to implement it yet (or had time to make a PR)

You mean that we add a dedicated boolean GPIO in addition to "position", "velocity",..? then we would have only one possible boolean-gpio-interface per gpio-tag. Or you'd like to add a boolean-gpio-interface to the joints?

Reply:
Let me clarify by using the [example 10](https://github.com/ros-controls/ros2_control_demos/pull/256/files#diff-8c788517881ed41ce5c891435bdabdf176f9e31b3848263fc04f17b7af5a4f18), more specifically

```
      <gpio name="flange_vacuum">
        <command_interface name="vacuum"/>
        <state_interface name="vacuum"/> <!-- Needed to know current state of the input -->
      </gpio>
```

It doesn't make much sense to me how the command and state interface names of this GPIO are `vacuum`. To me a command/state interface is the "physics" type of the value (position, velocity.... possibly voltage, temperature....).
`vacuum` is ambiguous; does the value represent the suction strength? or whether it's on or off?

What I'm proposing is adding in `hardware_interface_type_values.hpp` a HW_IF_BOOLEAN (or similarly named) to represent a common state/command interface not only for GPIOs, but also to joints and sensors.

I guess this leads to what you mentioned then:

> Supporting other datatypes (boolean, or even ROS msgs) is a repeating topic but no one made a sustainable suggestion on how to implement it yet (or had time to make a PR)

Reply:
I totally agree, the interface vacuum could be named more specifically.

But by just adding a boolean to [the list](https://github.com/ros-controls/ros2_control/blob/master/hardware_interface/include/hardware_interface/types/hardware_interface_type_values.hpp) would only allow one state- or command interface of this type per sensor/joint/GPIO [component](https://github.com/ros-controls/ros2_control/blob/d333922528ae009b3f30d31fe7ffcfe54edd45bb/hardware_interface/include/hardware_interface/hardware_info.hpp#L51) in this form:

      <gpio name="flange_vacuum">
        <command_interface name="boolean"/>
        <state_interface name="boolean"/>
      </gpio>